doc/TclString.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
.so man.macros
.TH TclString 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclString -- This class is used to create and modify Tcl string
objects. 

.SH EXTENDS
tcl.lang.InternalRep

.SH METHODS
.sp
static TclObject \fBnewInstance\fR(String \fIstrValue\fR)
.sp
static void \fBappend\fR(TclObject \fItobj\fR, String \fIstring\fR)

.SH ARGUMENTS
.AS TclObject "strValue" .
.AP TclObject tobj
TclObject to append a string to.  If the internal representation of
\fItobj\fR is not TclString, the method converts it to TclString.
.AP String strValue
Value of internal string in newly created TclObject.
.AP String string
String to append to the end of the string in the internal
representation of \fItobj\fR.
.BE

.SH DESCRIPTION
.TP
\fBnewInstance\fR
The \fBnewInstance\fR method creates and returns a new TclString whose internal
representation contains the string \fIstrValue\fR.
.TP
\fBappend\fR
The \fBappend\fR method concatenates \fIstring\fR to the end of the String
internally to \fItobj\fR.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_NewStringObj, Tcl_AppendToObj

.SH "SEE ALSO"
InternalRep, TclObject, TclList, TclIndex, TclDouble, TclInteger, TclBoolean, ReflectObject

.SH KEYWORDS
internal representation, object, object type, string, append


ChangeLog
2019-05-04  Fabrice Ducos <fabrice.ducos@gmail.com>

        * Support of latest versions of Tcl (at least Tcl 8.5
        and Tcl 8.6) and of modern JDK (10+);
        support of JDK 4- abandoned (use of auto-boxing);
        support of OSX systems added;
        details below

        * configuration (tcljava.m4, configure.in),
        and compilation (Makefile.in) templates updated for
        JDK 10+ (javah not available anymore).
        javah is still used when it is available (JDK 9-).
        * minor update of jtclsh.in (for OSX dynamic libraries)
        * many obsolescent constructs (mostly old-style boxing)
        removed in *.java under src/jacl/tcl/lang,
        src/tclblend/tcl/lang and src/tcljava/tcl/lang,
        in order to silence many warnings.
        (effect: support of JDK 4- abandoned)
        * src/native/javaInterp.c: 2010 implementation
        of Java_tcl_lang_Interp_initName() crashed on the
        example given at docs/website/getstart.html,
        because calling Tcl_GetNameOfExecutable
        before Tcl_FindExecutable gives rise to undefined
        behaviour, at least in latest versions of Tcl
        (crashes observed on several systems, at least with
        Tcl 8.6).
        Java_tcl_lang_Interp_initName() is now simplified and
        always calls Tcl_FindExecutable first.
        * all the occurences of interp->result have been
        replaced with Tcl_GetStringResult(interp).
        Accessing interp->result directly was deprecated in
        latest versions of the Tcl C API, and removed from
        Tcl 8.6 (another solution was to #define USE_INTERP_RESULT,
        but this would have been an easy-going, quick and dirty way)
        ref. https://www.tcl.tk/man/tcl/TclLib/Interp.htm

2019-05-01  Fabrice Ducos <fabrice.ducos@gmail.com>

        * tcljava resurrected
        * migration from a CVS to a GIT repository
        (with reposurgeon)

2010-02-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* tests/tcl/regexp.test: Add test case for
        escaped - character in regexp command.
        [Bug 2012389]

2010-02-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceAll):
	Fixup regsub -all logic so 5 more tests pass.
	The replaceAll() logic is broken because it depends
	on matching substrings instead of using an index
	into the larger string and Java's regexp library
	uses this to calculate line starts in multiline mode.

2010-02-18  Mo DeJong  <mdejong@users.sourceforge.net>

        Incorporate fix/tests from Tcl head on 2010-02-11,
        aka was Tcl [Bug 2826551, Patch 2948425].

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
        Fixup increment index on zero length match issue.
	* tests/tcl/regexp.test: Back port of new regexp tests,
        some tests marked with {endIndex} due to lack of impl.

2010-02-11  Mo DeJong  <mdejong@users.sourceforge.net>

	Patch 2946118

	* src/jacl/tcl/lang/ListCmd.java: Minor style reformat.
	* src/jacl/tcl/lang/Util.java (scanElement, convertElement):
	Remove crusty old code that treated backslash zero
	as the end of the line. That logic was removed from
	C Tcl long ago.
	* tests/inprogress/EventAdaptor.test: Fixup tests
	that depended on backslash zero quirks.
	* tests/tcl/list.test: Add test cases for backslash zero
	as argument to list command.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (match, _match,
        replaceFirst, replaceAll): Rewrite -start offset
        logic to correctly deal with odd edge cases.
	* tests/tcl/regexp.test: Add additional test cases
        that cover -start index edge cases.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Rewrite argument processing logic to avoid using
        IndexOutOfBoundsException to deal with an incorrect
        number of arguments.

2009-10-03  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fix regsub -all crash because offset was not getting
        advanced by 1 on a match of length zero.
        Fixup -start argument support with and without -all
        in the case where the offset was larger than the
        length of the string.
	* tests/tcl/regexp.test: Add additional test cases.
        Enable regsub test cases that were getting caught
        in an infinite loop.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        failing tests related to optional subgroups.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Minor paren fix, no function changes.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Rework
        logic related to match result groups.
	* tests/tcl/regexp.test: Add missing test cases for
        non-capturing groups and -inline results.

2009-09-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup use of MULTILINE flag.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for -linestop and -lineanchor arguments to regexp.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):  Fixup
        support for -linestop and -lineanchor arguments to regsub.
	* tests/tcl/regexp.test: Add missing test cases for
        -linestop and -lineanchor arguments.

2009-09-22  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup support for the empty string as the string
        argument to regsub by removing the MULTILINE flag when
        an empty string is detected.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for empty string as string argument.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Remove comments.
	* tests/tcl/regexp.test: Add additional test cases

tcljava/src/tcljava/tcl/lang/TclString.java
/*
 * TclList.java
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclString.java,v 1.11 2006/05/30 17:28:01 mdejong Exp $
 *
 */

package tcl.lang;

// This class implements the string object type in Tcl.

public class TclString implements InternalRep {

    // This dummy field is used as the internal rep for every
    // TclString that has not been modified via an append
    // operation. The most common case is for a TclString
    // to be created but never be appended to. This field
    // makes it possible to avoid allocating an internal
    // rep instance until a string is actually modified.

    private static TclString dummy = new TclString();

    // Used to perform "append" operations. After an append op,
    // sbuf.toString() will contain the latest value of the string and
    // tobj.stringRep will be set to null. This field is not private
    // since it will need to be accessed directly by Jacl's IO code.

    StringBuffer sbuf;

    private TclString() {
	sbuf = null;

	if (TclObject.saveObjRecords) {
	    String key = "TclString";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = 1;
	    } else {
	        num = num + 1;
	    }
	    TclObject.objRecordMap.put(key, num);
	}
    }

    private TclString(StringBuffer sb) {
	sbuf = sb;

	if (TclObject.saveObjRecords) {
	    String key = "TclString";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = 1;
	    } else {
	        num = num + 1;
	    }
	    TclObject.objRecordMap.put(key, num);
	}
    }

    /**
     * Returns a dupilcate of the current object.
     * @param obj the TclObject that contains this internalRep.
     */

    public InternalRep duplicate() {
	if (TclObject.saveObjRecords) {
	    String key = "TclString.duplicate()";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = 1;
	    } else {
	        num = num + 1;
	    }
	    TclObject.objRecordMap.put(key, num);
	}

	return dummy;
    }

    /**
     * Implement this no-op for the InternalRep interface.
     */

    public void dispose() {}

    /**
     * Called to query the string representation of the Tcl object. This
     * method is called only by TclObject.toString() when
     * TclObject.stringRep is null.
     *
     * @return the string representation of the Tcl object.
     */
    public String toString() {
	if (sbuf == null) {
	    return "";
	} else {
	    return sbuf.toString();
	}
    }

    /**
     * Create a new TclObject that has a string representation with
     * the given string value.
     */
    public static TclObject newInstance(String str) {
	return new TclObject(dummy, str);
    }

    /**
     * Create a new TclObject that makes use of the given StringBuffer
     * object. The passed in StringBuffer should not be modified after
     * it is passed to this method.
     */
    public static TclObject newInstance(StringBuffer sb) {
	return new TclObject(new TclString(sb));
    }

    static final TclObject newInstance(Object o) {
	return newInstance(o.toString());
    }

    /**
     * Create a TclObject with an internal TclString representation
     * whose initial value is a string with the single character.
     *
     * @param c initial value of the string.
     */

    static final TclObject newInstance(char c) {
	char charArray[] = new char[1];
	charArray[0] = c;
	return newInstance(new String(charArray));
    }

    /**
     * Called to convert the other object's internal rep to string.
     *
     * @param tobj the TclObject to convert to use the TclString internal rep.
     */
    private static void setStringFromAny(TclObject tobj) {
	// Create string rep if object did not have one already

	tobj.toString();

	// Change the type of the object to TclString.

	tobj.setInternalRep(dummy);

	if (TclObject.saveObjRecords) {
	    String key = "String -> TclString";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = 1;
	    } else {
	        num = num + 1;
	    }
	    TclObject.objRecordMap.put(key, num);
	}
    }

    /*
     * public static String get(TclObject tobj) {;}
     *
     * There is no "get" class method for TclString representations.
     * Use tobj.toString() instead.
     */


    /**
     * Appends a string to a TclObject object. This method is equivalent to
     * Tcl_AppendToObj() in Tcl 8.0.
     *
     * @param tobj the TclObject to append a string to.
     * @param string the string to append to the object.
     */
    public static final void append(TclObject tobj, String string) {
	if (!tobj.isStringType()) {
	    setStringFromAny(tobj);
	}

	TclString tstr = (TclString) tobj.getInternalRep();
	if (tstr == dummy) {
	    tstr = new TclString();
	    tobj.setInternalRep(tstr);
	}
	if (tstr.sbuf == null) {
	    tstr.sbuf = new StringBuffer(tobj.toString());
	}
	tobj.invalidateStringRep();
	tstr.sbuf.append(string);
    }

    /**
     * Appends an array of characters to a TclObject Object.
     * Tcl_AppendUnicodeToObj() in Tcl 8.0.
     *
     * @param tobj the TclObject to append a string to.
     * @param charArr array of characters.
     * @param offset index of first character to append.
     * @param length number of characters to append.
     */
    public static final void append(TclObject tobj,
            char[] charArr, int offset, int length) {
	if (!tobj.isStringType()) {
	    setStringFromAny(tobj);
	}

	TclString tstr = (TclString) tobj.getInternalRep();
	if (tstr == dummy) {
	    tstr = new TclString();
	    tobj.setInternalRep(tstr);
	}
	if (tstr.sbuf == null) {
	    tstr.sbuf = new StringBuffer(tobj.toString());
	}
	tobj.invalidateStringRep();
	tstr.sbuf.append(charArr, offset, length);
    }

    /**
     * Appends a TclObject to a TclObject. This method is equivalent to
     * Tcl_AppendToObj() in Tcl 8.0.
     *
     * The type of the TclObject will be a TclString that contains the
     * string value:
     *		tobj.toString() + tobj2.toString();
     */
    static final void append(TclObject tobj, TclObject tobj2) {
	append(tobj, tobj2.toString());
    }

    /**
     * Appends the String values of multiple TclObject's to a
     * TclObject. This is an optimized implementation that
     * will measure the length of each string and expand the
     * capacity as needed to limit reallocations.
     *
     * @param tobj the TclObject to append elements to.
     * @param objv array containing elements to append.
     * @param startIdx index to start appending values from
     * @param endIdx index to stop appending values at
     */

    static final void append(TclObject tobj,
	    TclObject[] objv,
            final int startIdx, final int endIdx) {
	if (!tobj.isStringType()) {
	    setStringFromAny(tobj);
	}

	TclString tstr = (TclString) tobj.getInternalRep();
	if (tstr == dummy) {
	    tstr = new TclString();
	    tobj.setInternalRep(tstr);
	}
	if (tstr.sbuf == null) {
	    tstr.sbuf = new StringBuffer(tobj.toString());
	}
	StringBuffer sb = tstr.sbuf;
	int currentLen = tstr.sbuf.length();

	tobj.invalidateStringRep();

	for (int i=0; i < endIdx; i++) {
	    currentLen += objv[i].toString().length();
	}
	// Large enough to holds all bytes, plus a little extra
	if (currentLen > (1024 * 10)) {
	    currentLen += (currentLen / 10);
	} else {
	    currentLen += (currentLen / 4);
	}
	sb.ensureCapacity(currentLen);
	for (int i=0; i < endIdx; i++) {
	    sb.append(objv[i].toString());
	}
    }

    /**
     * This procedure clears out an existing TclObject so
     * that it has a string representation of "". This
     * method is used only in the IO layer.
     */

    public static void empty(TclObject tobj) {
	if (!tobj.isStringType()) {
	    setStringFromAny(tobj);
	}

	TclString tstr = (TclString) tobj.getInternalRep();
	if (tstr == dummy) {
	    tstr = new TclString();
	    tobj.setInternalRep(tstr);
	}
	if (tstr.sbuf == null) {
	    tstr.sbuf = new StringBuffer();
	} else {
	    tstr.sbuf.setLength(0);
	}
	tobj.invalidateStringRep();
    }
}

